benchmark/bls.cpp

```diff
@@ -0,0 +1,417 @@
+#include <benchmark.hpp>
+#include <eosio/chain/apply_context.hpp>
+#include <eosio/chain/webassembly/interface.hpp>
+#include <eosio/testing/tester.hpp>
+#include <test_contracts.hpp>
+#include <bls12-381/bls12-381.hpp>
+#include <random>
+
+using namespace eosio;
+using namespace eosio::chain;
+using namespace eosio::testing;
+using namespace bls12_381;
+
+// benchmark BLS host functions directly without relying on CDT wrappers.
+// to run a benchmarking session, in the build directory, type
+// benchmark/benchmark -f bls
+
+namespace eosio::chain {
+// placed in eosio::chain name space so that interface_in_benchmark
+// can access transaction_context's private member max_transaction_time_subjective.
+// interface_in_benchmark is a friend of transaction_context.
+
+// To benchmark host functions directly without CDT wrappers,
+// we need to contruct a eosio::chain::webassembly::interface object,
+// as host functions are implemented in eosio::chain::webassembly::interface
+struct interface_in_benchmark {
+   interface_in_benchmark() {
+      // prevent logging from messing up benchmark results display
+      fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::off);
+
+      // create a chain
+      fc::temp_directory tempdir;
+      auto conf_genesis = tester::default_config( tempdir );
+      auto& cfg = conf_genesis.second.initial_configuration;
+      // configure large cpu usgaes so expensive BLS functions like pairing
+      // can run a reasonable number of times within a trasaction time
+      cfg.max_block_cpu_usage        = 500'000;
+      cfg.max_transaction_cpu_usage  = 480'000;
```

Comment:
Seems like these could be set much much larger so you don't have to limit benchmarking to 480ms.

Reply:
I was thinking we should be close to real block and transaction cpu limits. We don't need that. After I changed to very large values, all expensive functions can be benchmarked without any issues. Thanks!